ambuda/xml.py

```diff
@@ -38,16 +38,18 @@ class Rule:
 
     #: The tag to apply to this element.
     tag: str
-    #: Attributes to apply to this element.
-    attrib: dict
+    #: Attributes to apply to this element. Can be specified
+    #: as a dict, or as a function that inputs the existing
+    #: element's attributes and outputs a dict of new attributes.
+    attrib: Union[dict, Callable]
     #: Text to insert before this element's `text` field.
     text_before: str
     #: Text to insert after this element's `tail` field.
     text_after: str
 
     def __call__(self, el: ET.Element):
         el.tag = self.tag
-        el.attrib = self.attrib
+        el.attrib = self.attrib(el.attrib) if callable(self.attrib) else self.attrib
```

Comment:
It doesn't feel right to overload `attrib` like this. Instead, let's make `attrib` only a `Callable` and update `elem` to use a new function:

```
def elem(tag, attrib=None, text_before="", text_after="") -> Rule:
    """Helper to rename an element and change its attributes."""
    return Rule(tag, lambda _: attrib or {}, text_before, text_after)
```

Then:

```
    "ref": elem("a", remap({'href': 'target'}),
```

with an appropriate implementation of `remap`.

Reply:
I'm not quite sure what you're thinking `copy` would do. Can you elaborate?

If `attrib` only takes a `Callable`, though, then wouldn't that make things much more verbose?

i.e., for the existing code:

```
"body": elem("li", {"class": "dict-entry"}),
```

You'd need to wrap `{"class": "dict-entry"}` in another function.

Reply:
> I'm not quite sure what you're thinking copy would do. Can you elaborate?

Sure, something like this (perhaps `overwrite` is a better name):

```
def overwrite(new_attrib):
  def inner(_):
    return new_attrib
  return inner

[...]
```

Then while `{"class": "dict-entry"}` does needs to be wrapped, that wrapping occurs in `elem`, so the high-level API is identical to what we have now (so, no verbosity for callers) while the underlying structure is based on simple `Callable`s without overloaded types (so, a simpler interface for `Rule`).